### PR TITLE
Add support for SHA-256

### PR DESCRIPTION
**Pull request by bk2204**
This series introduces support for SHA-256 into gitobj.  Git itself will likely learn support for SHA-256 in either 2.28 or 2.29, and we'll need to follow suit.

The series is split out into separate logical and independent patches.  It does introduce a breaking change to use a set of options for the object DB, so once it's merged we'll need to do a v2.0.0 release.  The `go.mod` and package paths are updated at the commit that introduces the breaking change.

Because many of the tests are very similar but just indented one more level, it may be helpful to review this series with whitespace ignored.  I've opted to use a set of integration tests to test the different algorithms because most of the tests for individual object types are not interested in the hash value, but rather various different parsing edge cases.

### CODE REVIEW

**Comment by bk2204** [comment]
Gentle ping, @git-lfs/core.  I know this is a big series and it will take some time to review (plus folks have had holidays), but it would be great to get in for a release this month if we can.

File: object_db.go [hunk 1]
```diff
@@ -383,3 +376,14 @@ func (o *ObjectDatabase) cleanup(f *os.File) {
 	f.Close()
 	os.Remove(f.Name())
 }
+
+func hasher(algo string) hash.Hash {
```
Comment (ttaylorr) [inline object_db.go:385]: I wouldn't be upset if this patch came first (since the introduction of this code in `Hasher()` and subsequent move here creates a little more reviewer burden, but I think that you could just as easily ignore this.

File: pack/index_v1.go [hunk 2]
```diff
@@ -13,19 +13,22 @@ type V1 struct {
 // Name implements IndexVersion.Name by returning the 20 byte SHA-1 object name
 // for the given entry at offset "at" in the v1 index file "idx".
 func (v *V1) Name(idx *Index, at int64) ([]byte, error) {
-	var sha [20]byte
-	if _, err := idx.readAt(sha[:], v1ShaOffset(at)); err != nil {
+	var sha [32]byte
```
[the remark this answers is not in the document]
Comment (chrisd8088) [reply pack/index_v1.go:16]: I was also able to do the following in the Go playground, so it might work to allocate arrays:
```
	var foo [sha256.Size]byte
	foo[0] = 1
```

File: pack/set.go [hunk 3]
```diff
@@ -139,7 +139,7 @@ func NewSetPacks(packs ...*Packfile) *Set {
 	}
 
 	return &Set{
-		m: m,
+		m:    m,
```
[the remark this answers is not in the document]
Comment (bk2204) [reply pack/set.go:142]: Yes, it was.  I just ran `goimports` to tidy it while I was there and I blindly committed all of the changes.

File: object_db.go (hunk 1 above)
Comment (bk2204) [reply object_db.go:385]: I've just squashed it into the appropriate patch earlier in the series.

File: object_db.go [hunk 4]
```diff
@@ -29,41 +32,72 @@ type ObjectDatabase struct {
...
+}
+
+type options struct {
+	alternates   string
+	objectFormat string
+}
+
+type Option func(*options)
+
+func Alternates(alternates string) Option {
```
[the remark this answers is not in the document]
Comment (bk2204) [reply object_db.go:47]: I agree, and I've added some.

File: object_db.go [hunk 5]
```diff
@@ -29,41 +32,72 @@ type ObjectDatabase struct {
...
+	// objectFormat is the object format (hash algorithm)
+	objectFormat string
+}
+
+type options struct {
+	alternates   string
+	objectFormat string
+}
+
+type Option func(*options)
```
[the remark this answers is not in the document]
Comment (bk2204) [reply object_db.go:45]: We unfortunately need to instantiate a variable in Git LFS itself with the appropriate type, since we need to build a slice of options that we can pass in.  I did intend to adopt this suggestion, but then realized that my generic instantiation function in Git LFS was not going to work if I did.

File: object_db.go [hunk 6]
```diff
@@ -29,41 +32,72 @@ type ObjectDatabase struct {
...
-		rw: rw,
-	}, nil
+	odb := &ObjectDatabase{
+		ro:           ro,
+		rw:           rw,
+		objectFormat: args.objectFormat,
+	}
+
+	if odb.Hasher() == nil {
+		return nil, fmt.Errorf("unknown object format %q", args.objectFormat)
```
[the remark this answers is not in the document]
Comment (bk2204) [reply object_db.go:103]: I've switched this to a fixed type instead, as you suggested.

File: pack/index_v2_test.go [hunk 7]
```diff
@@ -26,9 +28,6 @@ var (
 		0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3,
 		0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3, 0x3,
 	}
-	V2IndexSmallSha  = V2IndexNames[0:20]
-	V2IndexMediumSha = V2IndexNames[20:40]
-	V2IndexLargeSha  = V2IndexNames[40:60]
```
[the remark this answers is not in the document]
Comment (bk2204) [reply pack/index_v2_test.go:30]: Removed.

File: pack/index_v1_test.go [hunk 8]
```diff
@@ -69,12 +47,28 @@ func init() {
...
 
-	buf := make([]byte, 0, indexOffsetV1Start+(3*indexObjectEntryV1Width))
+	hashlen := hash.Size()
+	entrylen := hashlen + indexObjectCRCWidth
+	entries := make([]byte, entrylen*3)
+
+	for i := 0; i < 3; i++ {
+		// Set the first three bytes to 0 and the remainder to the same
+		// value.  That sets the CRC to 1..3 (as a 4-byte value) and each of the byte
+		// values in the object ID to the same value.
```
[the remark this answers is not in the document]
Comment (bk2204) [reply pack/index_v1_test.go:57]: I've taken your text and squashed in an attribution.

File: pack/index_v1.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bk2204) [reply pack/index_v1.go:16]: This is a great idea.  Fixed.
